ingestion-sink/src/main/java/com/mozilla/telemetry/ingestion/sink/transform/PubsubMessageToObjectNode.java

```diff
@@ -1,79 +1,536 @@
+          // include payload if present.
+          .filter(data -> !data.isEmpty()).ifPresent(data -> contents.put(FieldName.PAYLOAD, data));
+      // Also include client_id if present.
+      Optional.ofNullable(message.getAttributesOrDefault(Attribute.CLIENT_ID, null))
+          .ifPresent(clientId -> contents.put(Attribute.CLIENT_ID, clientId));
+      return contents;
+    }
   }
 
-  /**
-   * Like {@link #rawContents(PubsubMessage)}, but uses the nested metadata format of decoded pings.
-   *
-   * <p>We include most attributes as fields, but the nested metadata format does not include error
-   * attributes and only includes {@code metadata.uri} when document namespace is
-   * {@code "telemetry"}.
-   */
-  private static ObjectNode decodedContents(PubsubMessage message) {
-    ObjectNode contents = Json
-        .asObjectNode(AddMetadata.attributesToMetadataPayload(message.getAttributesMap()));
-    // bytes must be formatted as base64 encoded string.
-    Optional.of(BASE64_ENCODER.encodeToString(message.getData().toByteArray()))
-        // include payload if present.
-        .filter(data -> !data.isEmpty()).ifPresent(data -> contents.put(PAYLOAD, data));
-    // Also include client_id if present.
-    Optional.ofNullable(message.getAttributesOrDefault(Constant.Attribute.CLIENT_ID, null))
-        .ifPresent(clientId -> contents.put(Constant.Attribute.CLIENT_ID, clientId));
-    return contents;
+  public static class Payload extends PubsubMessageToObjectNode {
+
+    public static Payload of(List<String> strictSchemaDocTypes, String schemasLocation,
+        String schemaAliasesLocation, IOFunction<String, InputStream> open) {
+      return new Payload(strictSchemaDocTypes, schemasLocation, schemaAliasesLocation, open);
+    }
+
+    private final Cache<String, String> normalizedNameCache = CacheBuilder.newBuilder()
+        .maximumSize(50_000).build();
+    private final BigQuerySchemaStore schemaStore;
+    private final Predicate<PubsubMessage> strictSchema;
+
+    /**
+     * Transform {@link PubsubMessage} into {@link ObjectNode} in live table format.
+     */
+    private Payload(List<String> strictSchemaDocTypes, String schemasLocation,
+        String schemaAliasesLocation, IOFunction<String, InputStream> open) {
+      if (strictSchemaDocTypes == null) {
+        strictSchema = message -> false;
+      } else {
+        final HashSet<String> strictSchemaDocTypeSet = new HashSet<>(strictSchemaDocTypes);
+        strictSchema = message -> {
+          String namespace = message.getAttributesOrDefault(Attribute.DOCUMENT_NAMESPACE, "");
+          String docType = message.getAttributesOrDefault(Attribute.DOCUMENT_TYPE, "");
+          return strictSchemaDocTypeSet.contains(namespace + "/" + docType);
+        };
+      }
+      schemaStore = BigQuerySchemaStore.of(schemasLocation, schemaAliasesLocation, open);
+    }
+
+    /** measure rate of CoercedToInt. */
+    private void incrementCoercedToInt() {
```

Comment:
> I'm guessing the big problem here is that we need to manipulate TableRow objects in the one context and ObjectNode in the other. Do you also see that as the blocker? We should make sure we think through that problem and make sure there's no way around it.

I absolutely think we can share this code between the two. jackson's ObjectMapper can be used to convert the ObjectNode to a TableRow. It's inconvenient that PubsubMessage is different in beam, but we can also work around that. The only real reason I didn't make this shared is that it will take longer to write.

 I'll file an issue to come back and share the implementation.